### ISSUE TITLE
Pasting large(ish) images results in segfault in ImageAlterClut()

### ISSUE THREAD

**Issue by miiichael** [fontforge/fontforge#3760]
Initially encountered with the version in Debian buster (20170731), but also reproducable from latest git (5b5b1b9ecde86db046cf7746b0970faa3529957f as of writing).

gdb backtrace:
```(gdb) run
Starting program: /home/michael/src/fontforge/fontforgeexe/.libs/fontforge 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Copyright (c) 2000-2019. See AUTHORS for Contributors.
 License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
 with many parts BSD <http://fontforge.org/license.html>. Please read LICENSE.
 Based on sources from 16:30 UTC  4-Jul-2019-ML-D.
 Based on source from git with hash: 5b5b1b9ecde86db046cf7746b0970faa3529957f
Failed to open hotkey definition file: /home/michael/src/fontforge/fontforgeexe/share/fontforge/hotkeys/default
Failed to open resource file: /home/michael/src/fontforge/fontforgeexe/share/fontforge/pixmaps/resources
Recovering from /home/michael/.config/fontforge/autosave/auto0062d0-2.asfd...  Done
Read Error

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff763d37d in ImageAlterClut (image=0x0) at cvimages.c:875
875	    struct _GImage *base = image->list_len==0?image->u.image:image->u.images[0];
(gdb) bt
#0  0x00007ffff763d37d in ImageAlterClut (image=0x0) at cvimages.c:875
#1  0x00007ffff763d69b in SCAddScaleImage (sc=0x555555979a40, image=<optimized out>, doclear=1, layer=1) at cvimages.c:947
#2  0x00007ffff763e415 in SCCheckXClipboard (sc=0x555555979a40, layer=layer@entry=1, doclear=doclear@entry=1) at cvundoes.c:2163
#3  0x00007ffff7647559 in PasteIntoFV (fv=<optimized out>, pasteinto=0, trans=0x0) at cvundoes.c:3626
#4  0x00007ffff75072b0 in gmenu_mouse (event=0x7fffffffbce0, m=0x5555559e06b0) at gmenu.c:948
#5  0x00007ffff75072b0 in gmenu_mouse (m=<optimized out>, event=0x7fffffffbce0) at gmenu.c:842
#6  0x00007ffff75077fb in gmenu_eh (w=0x5555559e07c0, ge=0x7fffffffbce0) at gmenu.c:1389
#7  0x00007ffff752ece2 in dispatchEvent (gdisp=gdisp@entry=0x5555555ef470, event=<optimized out>, event@entry=0x7fffffffbe40) at gxdraw.c:3283
#8  0x00007ffff7530e25 in GXDrawEventLoop (gd=0x5555555ef470) at gxdraw.c:3391
#9  0x00007ffff7dffc0e in fontforge_main (argc=<optimized out>, argv=<optimized out>) at startui.c:1432
#10 0x00007ffff700c09b in __libc_start_main () at /lib/x86_64-linux-gnu/libc.so.6
#11 0x000055555555508a in _start ()
(gdb) quit
```

The [image](https://user-images.githubusercontent.com/4179353/60681698-d66fb580-9ec2-11e9-9e7a-f79f4376dd58.png) I attempted to paste is 396x456 pixels. While I expect this may become unwieldy if _every_ glyph contained a background image this large (on the other hand, I have 32GB RAM...), I don't think the program should crash. :stuck_out_tongue:

**Comment by ctrlcctrlv** [comment]
I really doubt this is due to the size of the image. I'm working on a font right now where most glyphs have a background that is 16MP or higher. I'll do some debugging and get back to you later today.

**Comment by miiichael** [comment]
I did do some hamfisted poking around in cvundoes.c last night, but only got as far as noticing that in the crashing case, the retrieved PNG clip was only 4 bytes, which made me wonder if an error path somewhere (in or around GImageRead_Png?) needs some love, especially in the presence of the vague and mysterious "Read Error" message.

**Comment by ctrlcctrlv** [comment]
To reproduce this bug, you must have the image in the X clipboard. Saving the file into the disk doesn't cause the segfault. With that in mind, pasting into the CharView causes the segfault as well as pasting into the FontView.

A fix for now is just to save the image to the disk and use _File&rarr;Import&hellip;_

**Comment by ctrlcctrlv** [comment]
```diff
diff --git a/fontforge/cvundoes.c b/fontforge/cvundoes.c
index 20e7bd02d..6a497e6fc 100644
--- a/fontforge/cvundoes.c
+++ b/fontforge/cvundoes.c
@@ -2146,15 +2146,19 @@ return;
 return;
 
     temp = GFileTmpfile();
+    FILE* temp2 = fopen("/tmp/output","w+");
     if ( temp!=NULL ) {
 	fwrite(paste,1,len,temp);
+	fwrite(paste,1,len,temp2);
 	rewind(temp);
+	rewind(temp2);
 	if ( type==4 || type==5 ) {	/* eps/ps */
 	    SCImportPSFile(sc,layer,temp,doclear,-1);
 	} else if ( type<=3 ) {
 	    SCImportSVG(sc,layer,NULL,paste,len,doclear);
 	} else {
 #ifndef _NO_LIBPNG
+        printf("Writing PNG\n");
 	    if ( type==6 )
 		image = GImageRead_Png(temp);
 	    else
```

shows that junk is received from `ClipboardRequest` and written into `/tmp/output` (and therefore is really not a PNG as FontForge claims):

```xxd
00000000: e7c6 0400 0000 0000                      ........
```

But if a known working image is used, such as this one:

![DEF3](https://user-images.githubusercontent.com/838783/60699124-23847380-9f25-11e9-877b-27deedcce9d8.png)

`/tmp/output` contains a PNG file as expected:

```
[fred@pc tmp]$ file /tmp/output
/tmp/output: PNG image data, 475 x 399, 8-bit/color RGBA, non-interlaced
```

# TL;DR: Something is wrong with the communication between X and FontForge

**Comment by jtanx** [comment]
There's nothing actually wrong with the protocol, it's just that neither the x11 nor the gdk backend actually supports receiving clipboard data when it's sent with the [INCR](https://www.x.org/releases/X11R7.6/doc/xorg-docs/specs/ICCCM/icccm.html#incr_properties) protocol:


```
22:00:49.189 DEBUG: _GGDKDraw_DispatchEvent (ggdkdraw.c:1211) - SEL NOTIFY Sel(0x45 CLIPBOARD) Target(0x8b image/png) Property(0x72 GDK_SELECTION)
22:00:49.194 DEBUG: GGDKDrawRequestSelection (ggdkdraw.c:1956) - RECEIVED TYPE 0xa3(INCR)
```

**Comment by ctrlcctrlv** [comment]
So we either have to implement that or throw a warning it seems. Implementation isn't exactly "bite sized" but may not be that difficult; throwing a warning to prevent the crash is easy.
